### ISSUE TITLE
Declaration does not compile on ts 0.9.1 due to overloaded methods with only differing return type

### ISSUE THREAD

**Issue by kirki** [brian428/ext-typescript-generator#1]
In typescript 0.9.1 a bug was fixed that allowed to overload methods ony be return type even the spec said it is an error (http://typescript.codeplex.com/workitem/1373).

This pattern used in the ext-typescript-generator quite often to support the ambigous return types of many extjs methods. So the generated declarations file fails to compile with over 100 errors.

To fix this with 0.9.1 I would see 3 approaches to use when multiple return types would be needed:
- use generics to define return type
- use common super class of all used return types
- just use "any" as return type

The generic approach could look like this:

``` javascript
declare module Ext.container
{
   // artificial interface to form supercass of generic type for following "add" method
   interface IAbstractContainer_Add_ReturnType extends Ext.IComponent, Array<Ext.IComponent> { }

   interface IAbstractContainer
   {
      // instead of these 2 delcarations:
      //add?( component?:any ): Ext.IComponent[];
      //add?( component?:any ): Ext.IComponent;  

      // use this declaration:
      add? <T extends IAbstractContainer_Add_ReturnType>(component?: any): T;
   }
}
```

---

This is correct typescript code but unfortunately brings no real benefit. The typescript compiler is not that strict when considering generic return types.

So the following examples all work:

``` javascript
var c: Ext.container.IAbstractContainer;
var r1: Ext.IComponent = c.add('test1');
var r2: Ext.IComponent[] = c.add('test2');
var r3: string = c.add('test3');  // oops - also ok. tsc is not that strict here
```

So I suppose the best and most simple approach would be the use of "any" instead.

**Comment by brian428** [comment]
Are you using the version at definitelytyped? I've been building against 0.9.1 for several weeks without any problems. I'll have to make sure the version I have here is up to date, but the one at https://github.com/borisyankov/DefinitelyTyped/tree/master/extjs should work with 0.9.1.

**Comment by kirki** [comment]
I am quite sure and checked it once again. The latest version in DefinitelyTyped (commit dd35f69) does not work with the latest typescript version 0.9.1.
You committed some changes on 14th of July with a comment that you made it work for ts 0.9.1. But this was weeks before the 0.9.1 actually was released. Maybe you mixed it with the previous ts version 0.9.0.1?

**Comment by brian428** [comment]
Yep, you're right, I was indeed on 0.9.0.1. I'll update and see what I have
to do now to get things working again. I'm happy that they're updating TS
rapidly, but it seems like almost every update breaks something (or "fixes"
something that was working and wasn't supposed to). I hope things stabilize
and get a bit less jarring, because keeping up with the rate and impact of
the updates is difficult.

I really hate to switch the return types to "any", because having the
return types specified is part of the benefit. I get that it is weird to
have a signature that differs only by return type, but it's also not THAT
insane (different mixes of args can return different types), and it doesn't
really AFFECT anything, so for them to just nuke the option seems a bit
overzealous. But one way or the other I'll get it working.

**Comment by brian428** [comment]
Got things updated and working with 0.9.1. Pushed here and pull request pending at definitelytyped. Thanks for the heads up!

**Comment by brian428** [comment]
Actually just reopening to remind myself to do the same for the Touch definition.
